### PR TITLE
RESPA-277 | Add a new feature that allows cancelling reservations with predefined and custom reasons

### PR DESCRIPTION
**Pull request by tommimanbytes**
Add a new feature that allows cancelling reservations with predefined and custom reasons.

Reservation cancellation reason categories are first defined in Django Admin in all three languages. Cancellation reason categories all includes a predefined description text which is then added to cancellation / declination mails if reservation gets cancelled. Cancellation reasons are also shown to user in UI. The user which cancels the reservation can also type a custom cancellation message to reservation in addition to categorys predefined one.

Usage example

1. Define reservation cancellation reason categories in Django Admin
2. Cancel a reservation using PATCH HTTP-method with one of the defined categories and custom message: 
```json
{
  "state": "cancelled",
  "cancel_reason": {
    "description": "This reservation has been cancelled due to x and y",
    "category_id": 1
  }
}
```
3. Show cancel reason in UI to users and staff. Representation of cancel reason in reservation object in API:
```json
"cancel_reason": {
    "category": {
      "id": 1,
      "reservation_type": "confirmed",
      "name": {
        "fi": "grtre",
        "en": "gregr",
        "sv": "greg"
      },
      "description": {
        "fi": "regerg",
        "en": "egerg",
        "sv": "ergre"
      }
    },
    "description": "This reservation has been cancelled due to x and y",
  }
```

### CODE REVIEW

File: openapi.yaml [hunk 1]
```diff
@@ -1496,3 +1561,53 @@ components:
...
+          description: "Type of reservation for which this cancel reason category is used. Choices: confirmed, requested, own."
+        name:
+          type: object
+          properties:
+            sv:
+              type: string
+              description: Swedish name for the cancel reason category
+            fi:
+              type: string
+              description: Swedish name for the cancel reason category
```
Comment (indigane) [inline openapi.yaml:1598]: Swedish in all fields.

File: resources/api/reservation.py [hunk 2]
```diff
@@ -167,7 +193,7 @@ def validate(self, data):
         if not resource.can_make_reservations(request_user):
             raise PermissionDenied(_('You are not allowed to make reservations in this resource.'))
 
-        if data['end'] < timezone.now():
+        if data.get('end', False) and data['end'] < timezone.now():
```
Comment (indigane) [inline resources/api/reservation.py:196]: `if 'end' in data` is more explicit for testing key existence than `data.get()`. Mostly matters for readability. I had to think for a bit why `.get` is used before I understood this line.

File: openapi.yaml (hunk 1 above)
Comment (tommimanbytes) [reply openapi.yaml]: Fix'd

File: resources/api/reservation.py (hunk 2 above)
Comment (tommimanbytes) [reply resources/api/reservation.py]: Agreed, fix'd

**Pull request merged**